presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/type/TimestampRepresentations.java

```diff
@@ -0,0 +1,97 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.base.type;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.Int96ArrayBlock;
+import io.prestosql.spi.block.LongArrayBlock;
+import io.prestosql.spi.type.LongTimestamp;
+import io.prestosql.spi.type.TimestampType;
+import org.joda.time.DateTimeZone;
+
+import static io.prestosql.spi.type.TimestampType.MAX_SHORT_PRECISION;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP_NANOS;
+import static io.prestosql.spi.type.Timestamps.MICROSECONDS_PER_MILLISECOND;
+import static io.prestosql.spi.type.Timestamps.MICROSECONDS_PER_SECOND;
+import static io.prestosql.spi.type.Timestamps.PICOSECONDS_PER_MICROSECOND;
+import static io.prestosql.spi.type.Timestamps.epochMicrosToMillisWithRounding;
+import static io.prestosql.spi.type.Timestamps.round;
+import static io.prestosql.spi.type.Timestamps.roundDiv;
+import static java.lang.Math.multiplyExact;
+import static java.util.Objects.requireNonNull;
+
+public final class TimestampRepresentations
+{
+    private TimestampRepresentations() {}
+
+    public static PrestoTimestampEncoder<? extends Comparable<?>> createTimestampEncoder(TimestampType type, DateTimeZone timeZone)
+    {
+        requireNonNull(type, "type is null");
+        requireNonNull(timeZone, "timeZoneKey is null");
+
+        if (type.isShort()) {
+            return new ShortTimestampEncoder(type, timeZone);
+        }
+        return new LongTimestampEncoder(type, timeZone);
+    }
+
+    // copied from io.prestosql.type.DateTimes
+    static LongTimestamp longTimestamp(long epochSecond, long fractionInPicos)
+    {
+        return new LongTimestamp(
+                multiplyExact(epochSecond, MICROSECONDS_PER_SECOND) + fractionInPicos / PICOSECONDS_PER_MICROSECOND,
+                (int) (fractionInPicos % PICOSECONDS_PER_MICROSECOND));
+    }
+
+    /**
+     * Provide a canonical representation of the timestamp at the given position in the block with the precision
+     * of the target type.  Uses {@link LongTimestamp}, since it can represent any valid precision.  This is
+     * effectively a cast to a different precision timestamp and is useful in CTAS, when the engine won't do the
+     * cast for us.
+     */
+    public static LongTimestamp longTimestampRepresentation(Block block, int position, TimestampType type)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        if (block instanceof LongArrayBlock) {
```

Comment:
the conditions should be based on type.precision

the actual block representation can be many different things (like RLEBlock, DictionaryBlock, LazyBlock, etc)

Reply:
`type` represents the destination.  That does not tell us anything about the source precision.  What's a good mechanism to figure out whether we're dealing with long or short timestamps?